homeassistant/components/elkm1/config_flow.py

```diff
@@ -87,60 +105,214 @@ def _make_url_from_data(data):
+            self._discovered_device
+        )
+        return await self.async_step_discovered_connection()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
         """Handle the initial step."""
+        if user_input is not None:
+            if mac := user_input[CONF_DEVICE]:
+                await self.async_set_unique_id(mac, raise_on_progress=False)
+                self._discovered_device = self._discovered_devices[mac]
+                return await self.async_step_discovered_connection()
+            return await self.async_step_manual_connection()
+
+        current_unique_ids = self._async_current_ids()
+        current_hosts = {
+            urlparse(entry.data[CONF_HOST]).hostname
+            for entry in self._async_current_entries(include_ignore=False)
+        }
+        discovered_devices = await async_discover_devices(
+            self.hass, DISCOVER_SCAN_TIMEOUT
+        )
+        self._discovered_devices = {
+            dr.format_mac(device.mac_address): device for device in discovered_devices
+        }
+        devices_name: dict[str | None, str] = {
+            mac: f"{_short_mac(device.mac_address)} ({device.ip_address})"
+            for mac, device in self._discovered_devices.items()
+            if mac not in current_unique_ids and device.ip_address not in current_hosts
+        }
+        if not devices_name:
+            return await self.async_step_manual_connection()
+        devices_name[None] = "Manual Entry"
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema({vol.Required(CONF_DEVICE): vol.In(devices_name)}),
+        )
+
+    async def _async_create_or_error(
+        self, user_input: dict[str, Any], importing: bool
+    ) -> tuple[dict[str, str] | None, FlowResult | None]:
+        """Try to connect and create the entry or error."""
+        if self._url_already_configured(_make_url_from_data(user_input)):
+            return None, self.async_abort(reason="address_already_configured")
+
+        try:
+            info = await validate_input(user_input, self.unique_id)
+        except asyncio.TimeoutError:
+            return {CONF_HOST: "cannot_connect"}, None
+        except InvalidAuth:
+            return {CONF_PASSWORD: "invalid_auth"}, None
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            return {"base": "unknown"}, None
+
+        if importing:
+            return None, self.async_create_entry(title=info["title"], data=user_input)
+
+        return None, self.async_create_entry(
+            title=info["title"],
+            data={
+                CONF_HOST: info[CONF_HOST],
+                CONF_USERNAME: user_input[CONF_USERNAME],
+                CONF_PASSWORD: user_input[CONF_PASSWORD],
+                CONF_AUTO_CONFIGURE: True,
+                CONF_TEMPERATURE_UNIT: user_input[CONF_TEMPERATURE_UNIT],
```

Comment:
I don't like that we have to ask the user this.  The device added it to the protocol about 7 years ago so I'm going to see if we can get rid of this in another turn (not this PR because its already bigger than I had hoped)

Reply:
For temperature, the reason is that it is not discoverable what the ElkM1 is configured for (if I recall).

For the auto configure that might be something to collaborate on with the community. I like auto configure, mostly.  It doesn't work fully though since it is done through a heuristic (looking at when the name of a ElkM1 element is not the "default" name). 

I have toyed with being able to use auto configure with a limited option set to turn things on or off. For example, I use UPB lighting. Both Elk and UPB integrations support it, but I don't want any of the Elk lighting bits.

I'd love to remove the YAML config if we could get the right way to enable/disable some of the Elk elements.

Reply:
It looks like it was added to the `UA` response in 2005

```
Rev. 1.52_12/9/05 Added ‘F’ or ‘C’ to UA command for Fahrenheit or Celcius temperature mode.
Added “az” and “AZ” alarm by zone commands
```

https://elk-manuals.s3.amazonaws.com/ELK-M1_RS232_PROTOCOL.Ver+1.86.pdf
```
4.36.2 Reply - Valid User Code Areas (UA)
19 – Length as ASCII hex
UA – Returned valid user code areas
DDDDDD – 6 bytes of user code data that was requested. 
M1 Control RS-232 ASCII String Protocol Page 41 of 58 Rev. 1.86 Mar 6, 2017
NN – 8 bits of Hex Area Mask Data, using two ASCII Hex bytes,
0-9,A-F. Bit 0 = Area 1, code valid if set.
UUUUUUUU – 8 bytes diagnostic data, factory use only.
N – Number of digits in user code. “4” or “6” digits, Ver 4.2.8
L – Type of code, “1”= User, “2”= Master, “3” = Installer,
“4” = ELKRP, Ver 4.3.6 or later
T – F=Fahrenheit temperature mode, C=Celcius temperature mode.
M1 Ver. 4.3.9 or later
00 – future use
CC – Checksum
```

Reply:
> I'd love to remove the YAML config if we could get the right way to enable/disable some of the Elk elements.

The strategy Home Assistant has adopted for this is to create entities for everything and let the user disable the ones they don't want since entities can now be disabled in the UI.  That might be a bit overkill in this case since elk create a LOT of entities

Reply:
Agree on the entity creation/disabling. Doesn't work too well in this case. I wonder if there is something in between. 

WRT to UA, I remember there being something funky that it would not work - there were discussion back in the day. My old brain doesn't remember that far back. I'll poke away at it and see if I turn up anything. It wouldn't be hard to implement that UA handling.